Mathlib/Order/WellFoundedSet.lean

```diff
@@ -814,6 +815,97 @@ theorem partiallyWellOrderedOn_sublistForall₂ (r : α → α → Prop) [IsRefl
       exact List.SublistForall₂.cons (hg _ _ (le_of_lt mn)) hmn
 #align set.partially_well_ordered_on.partially_well_ordered_on_sublist_forall₂ Set.PartiallyWellOrderedOn.partiallyWellOrderedOn_sublistForall₂
 
+theorem subsetProdLex [PartialOrder α] [Preorder β] {s : Set (α ×ₗ β)}
+    (hα : ((fun (x : α ×ₗ β) => x.1)'' s).IsPWO)
+    (hβ : ∀ a, {y | (a, y) ∈ s}.IsPWO) : s.IsPWO := by
+  intro f hf
+  rw [isPWO_iff_exists_monotone_subseq] at hα
+  have hg : ∃ (g : (ℕ ↪o ℕ)), Monotone fun n => (f (g n)).1 :=
+    hα (fun n => (f n).1) (fun k => mem_image_of_mem (fun x => x.1) (hf k))
+  let g : (ℕ ↪o ℕ) := Exists.choose hg
+  have hhg : ∀ n, (f (g 0)).1 ≤ (f (g n)).1 := fun n => hg.choose_spec <| Nat.zero_le n
+  by_cases hc : ∃ n, (f (g 0)).1 < (f (g n)).1
+  · use (g 0), (g (Exists.choose hc))
+    let n := Exists.choose hc
+    constructor
+    · by_contra hx
+      rw [Nat.not_lt] at hx
+      have hhc : (f (g 0)).1 < (f (g n)).1 := hc.choose_spec
+      have hhg : g n = g 0 := by simp_all only [OrderEmbedding.le_iff_le, nonpos_iff_eq_zero]
+      simp_all [hhg]
+    · refine (Prod.Lex.le_iff (f (g 0)) _).mpr ?_
+      left
+      exact hc.choose_spec
+  · have hhc : ∀ n, (f (g 0)).1 = (f (g n)).1 := by
+      intro n
+      rw [@not_exists] at hc
+      exact (LE.le.not_lt_iff_eq (hhg n)).mp (hc n)
+    specialize hβ (f (g 0)).1
+    rw [isPWO_iff_exists_monotone_subseq] at hβ
+    specialize hβ fun n => (f (g n)).2
+    have hfg : ∀ (n : ℕ), ((f (g 0)).1, (f (g n)).2) ∈ s :=
+      fun n => mem_of_eq_of_mem (congrFun (congrArg Prod.mk (hhc n)) (f ((Exists.choose hg) n)).2)
+      (hf ((Exists.choose hg) n))
+    specialize hβ fun n => hfg n
+    let g' := Exists.choose hβ
+    use (g (g' 0)), (g (g' 1))
+    constructor
+    · simp only [OrderEmbedding.lt_iff_lt, zero_lt_one]
+    · have h: Monotone fun n => (f (g (g' n))).2 := hβ.choose_spec
+      refine (Prod.Lex.le_iff (f (g (g' 0))) (f (g (g' 1)))).mpr ?_
+      right
+      constructor
+      · exact (hhc (g' 0)).symm.trans (hhc (g' 1))
+      · exact h zero_le_one
+
+theorem imageProdLex [PartialOrder α] [Preorder β] {s : Set (α ×ₗ β)}
+    (hαβ : s.IsPWO) : ((fun (x : α ×ₗ β) => x.1)'' s).IsPWO :=
+  IsPWO.image_of_monotone hαβ Prod.Lex.monotone_fst
+
+theorem fiberProdLex [PartialOrder α] [Preorder β] {s : Set (α ×ₗ β)}
+    (hαβ : s.IsPWO) (a : α) : {y | (a, y) ∈ s}.IsPWO := by
```

Comment:
```suggestion
    (hαβ : s.IsPWO) (a : α) : {y | toLex (a, y) ∈ s}.IsPWO := by
```

Reply:
This probably applies in quite a few places; when using `(_, _)` you need `toLex`, and when using `.1` or `.2` on `α ×ₗ β` you should have `ofLex`.